### PR TITLE
Add multi select feature on TOC filter and group entries based on headers

### LINKED ISSUE #1114: Entries QA: Group entries based on headers (ToC)

### LINKED ISSUE #1175: Entries QA: Multi-selection in table of contents

### PR DESCRIPTION
**Pull request by samshara**
fixes https://github.com/the-deep/client/issues/1175
fixes https://github.com/the-deep/client/issues/1114

Depends on: https://github.com/the-deep/server/pull/572

### CODE REVIEW

File: src/components/TableOfContents/index.tsx [hunk 1]
```diff
@@ -62,7 +62,7 @@ function ToCItem<T, K extends string | number>(props: ToCItemProps<T, K>) {
         [],
     );
 
-    const isSelected = id === value?.id;
+    const isSelected = value.some(v => v.id === id);
```
Comment (AdityaKhatri) [inline src/components/TableOfContents/index.tsx]: Shouldn't we use keySelector?

File: src/views/Entries/QualityControl/index.tsx [hunk 2]
```diff
@@ -159,18 +158,23 @@ function QualityControl(props: Props) {
...
     const handleEntryDelete = React.useCallback((entryId) => {
         setDeletedEntries(oldDeletedEntries => ({ ...oldDeletedEntries, [entryId]: true }));
     }, [setDeletedEntries]);
 
-    const handleSelection = useCallback(value => (
-        selected && selected.id === value.id ?
-            setSelection({ matrixKey: undefined }) : setSelection({ matrixKey: value })
-    ), [selected, setSelection]);
+    const handleSelection = useCallback((value) => {
+        const isSelected = tocFilters.some(s => s.id === value.id);
```
Comment (AdityaKhatri) [inline src/views/Entries/QualityControl/index.tsx]: Use KeySelector

File: src/views/Entries/QualityControl/index.tsx [hunk 3]
```diff
@@ -159,18 +158,23 @@ function QualityControl(props: Props) {
...
+        if (isSelected) {
+            const newSelection = tocFilters.filter(s => s.id !== value.id);
```
Comment (AdityaKhatri) [inline src/views/Entries/QualityControl/index.tsx]: Use KeySelector

**Review by tnagorra — APPROVED** [review]

File: src/views/Entries/QualityControl/index.tsx [hunk 4]
```diff
@@ -229,17 +233,21 @@ function QualityControl(props: Props) {
         setDeletedEntries(oldDeletedEntries => ({ ...oldDeletedEntries, [entryId]: true }));
     }, [setDeletedEntries, getEntriesWithStats]);
 
-    const handleSelection = useCallback(value => (
-        selected && selected.id === value.id ?
-            setSelection({ matrixKey: undefined }) : setSelection({ matrixKey: value })
-    ), [selected, setSelection]);
+    const handleSelection = useCallback((value) => {
+        const isSelected = tocFilters.some(s => idSelector(s) === idSelector(value));
+        if (isSelected) {
```
Comment (AdityaKhatri) [inline src/views/Entries/QualityControl/index.tsx]: I think we should handle this from within the component. Thoughts @samshara @tnagorra ?

**Review by AdityaKhatri — CHANGES_REQUESTED** [review]
Pending conversations.

File: src/components/TableOfContents/index.tsx [hunk 5]
```diff
@@ -72,7 +72,7 @@ function ToCItem<T, K extends string | number>(props: ToCItemProps<T, K>) {
         [],
     );
 
-    const isSelected = id === value?.id;
+    const isSelected = value.some(v => idSelector(v) === id);
```
Comment (tnagorra) [inline src/components/TableOfContents/index.tsx]: let's memoize this

File: src/views/Entries/QualityControl/index.tsx (hunk 4 above)
Comment (tnagorra) [reply src/views/Entries/QualityControl/index.tsx]: yes

File: src/views/Entries/QualityControl/index.tsx [hunk 6]
```diff
@@ -259,6 +281,10 @@ function QualityControl(props: Props) {
         handleVerificationChange,
     ]);
 
+    const tocFilterRendererParams = (_: string, data: MatrixKeyId) => ({
+        value: labelSelector(data),
+    });
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/index.tsx]: Let' use useCallback?

File: src/components/TableOfContents/index.tsx [hunk 7]
```diff
@@ -56,11 +62,27 @@ function ToCItem<T, K extends string | number>(props: ToCItemProps<T, K>) {
...
+                    onChange([]);
+                }
+            } else if (multiple) {
+                onChange([
+                    ...value,
+                    option,
+                ]);
+            } else {
+                onChange([option]);
             }
```
Comment (tnagorra) [inline src/components/TableOfContents/index.tsx]:
```suggestion
            } else {
                onChange(multiple ? [...value, option] : [option]);
            }
```

File: src/components/TableOfContents/index.tsx [hunk 8]
```diff
@@ -56,11 +62,27 @@ function ToCItem<T, K extends string | number>(props: ToCItemProps<T, K>) {
...
+                return;
+            }
+            const isSelected = value.some(s => idSelector(s) === idSelector(option));
+            if (isSelected) {
+                if (multiple) {
+                    const newSelection = value.filter(s => idSelector(s) !== idSelector(option));
+                    onChange(newSelection);
+                } else {
+                    onChange([]);
+                }
```
Comment (tnagorra) [inline src/components/TableOfContents/index.tsx:75]: We can use ternary operator here as well
it's more readable that way. but we can also not do it for this case.

**Review by tnagorra — APPROVED** [review]

**Pull request merged**